### ISSUE TITLE
SetRoundMode should not be used at all ;)

### ISSUE THREAD

**Issue by tweibert** [AngusJohnson/Image32#45]
Hi,

in an earlier issue (#41) I pointed out that a try/finally block was missing in the `Rasterize` procedure of `Img32.Draw.pas`, because there are situations where the rounding mode (altered with `SetRoundMode`) is not reset properly.

Even with the fix apply, I kept seeing incorrect rounding modes, in particular in background threads, so I did a bit more research.

TL;DR: `SetRoundMode` should not be used in a multi-threaded application, because it is not thread-safe at all.

The answer to this SO question by David Heffernan is a good starting point to understand the issue: https://stackoverflow.com/questions/39684161/why-an-application-starts-with-fpu-control-word-different-than-default8087cw

(The SO question deals with floating point exceptions, but these are also controlled by bits/flags in the 8087 control word, just like the current rounding mode.)

In a nutshell, using `SetRoundMode` (or  more generally, any of Delphi's function to change the FPU flags, like `Set8087CW`) will store the current state in `Default8087CW`, which is a global variable in `System.pas`. And which is also used to initialize the FPU for any new thread being started (in `_FpuInit` in `System.pas`).

So when you temporarily change the rounding mode to `rmDown` via `SetRoundMode`, and a thread is started before you change it back to `rmNearest`, the thread will be initialized and run in `rmDown` mode. Here is a small console application that demostrates the problem: https://github.com/tweibert/RoundModeNotThreadSafe

`Img32.Draw.pas` does not use any threading, obviously, but it could be the case that there is an existing background thread that spawns another background thread, while the main thread is in the middle of executing the `Rasterize` function. Then the second background thread would suffer from this effect.

Now what is the solution? The bottom line is, the Delphi version of `SetRoundMode` should not be used. Instead, one could either:

- create a thread-safe version of `SetRoundMode` that only changes the FPU flags for the current thread using assembler code, and does not mess around with the `Default8087CW`. Of course, this would be platform-specific (I think only Win32 uses the FPU). FastMM4 is using this approach: https://github.com/pleriche/FastMM4/blob/8ee641096e4db76fbe9974ec454882601fd0a407/FullDebugMode%20DLL/FastMM_FullDebugMode.dpr#L220

or

- Do not use `Round()` at all but rather `Trunc()`. I saw comments in the code saying that `Trunc()` was rather slow, and also some (commented out) more efficient `__Trunc()` function that was used in earlier versions of Image32. As a quick fix, I re-enabled that function now, and changed all calls from `Round()` to `__Trunc()` in `Img32.Draw.pas`.

Thanks for looking into this.

Torben

EDIT: Regarding the former possible solution (thread-safe version of `SetRoundMode`), the following code could be used:

````delphi
procedure Set8087CWThreadSafe(ANewCW: Word);
var
  L8087CW: Word;
asm
  mov L8087CW, ANewCW
  fnclex
  fldcw L8087CW
end;

function SetRoundMode(const RoundMode: TRoundingMode): TRoundingMode;
var
  CtlWord: Word;
begin
  CtlWord := Get8087CW;
  Set8087CWThreadSafe((CtlWord and $F3FF) or (Ord(RoundMode) shl 10));
  Result := TFPURoundingMode((CtlWord shr 10) and 3);
end;
````

This is for 32-bit only. For x64, Delphi uses the MMX extensions instead. And other platforms? No idea, honestly.

**Comment by AngusJohnson** [comment]
Hi Torben.
Thank you for your very well researched post. 
I will certainly investigate and make changes.
However, currently I'm holidaying and unlikely to do much about this for another week or two.

**Comment by tweibert** [comment]
Thanks, and no hurries - as I said, I already found a workaround for me.

I was quite a bit surprised to learn that this `Set8087CW` thing is a well-known flaw in the RTL, has been around for years, but Embarcadero is not going to do anything about it. 

Enjoy your time off!

**Comment by AngusJohnson** [comment]
I've just retested the rasterizer using `Round` (not threadsafe), and `Trunc` and `__Trunc`.

Both Round (combined with SetRoundMode) and Trunc seem to be equally efficient now (ie rendering takes the same amount of time) and they are about 20% faster than __Trunc (which has been discarded).
I've added in a (likely temporary) compiler switch `UseTrunc` at the top of Img32.Draw. This is enabled by default since it means the rasterizer is now threadsafe. Unless unexpected problems appear, I'll remove the Round option in a later revision.

**Comment by tweibert** [comment]
@tomwiel Please see my original post and the links in it.

Delphi's `Set8087CW()` function does not only store the new value in the FPU register, but also in a global variable `Default8087CW`, which is also used as the initial value for any new thread started. So altering the value in one thread will have impact on other threads. In other words, it is not thread safe.

**Comment by tomwiel** [comment]
The solution is simple: EVERY thread should explicitly define the FPU register controlword 
(eg. by setRoundMode() and more), thus the initial thread default (from unsafe Default8087CW) will be replaced.

**Comment by tweibert** [comment]
No, that would not be a solution in my opinion, rather a workaround. i don't want to do that for every single thread. And how can I make sure that it is done properly in each and every 3rd party component I am using?

`Set8087CW` is not threadsafe. By using it, your code becomes non-threadsafe as well, obivously. So if @AngusJohnson wants Image32 to be threadsafe, the only solution is to avoid using this function altogether. Or put a big "we are not threadsafe" warning sign on the front page 😉

**Comment by tomwiel** [comment]
Set8087CW() is used in SetRoundMode(), SetPrecisionMode(), SetExceptionMask() and many more.
These calls cannot be forbidden, and they all modify the unprotected Default8087CW.  So there are MANY writers (Image32 is just one of them). But Default8087CW is only read in a very specific situation: when the thread is reset to a default value. 

The conclusion is: After thread reset (init), Programmers should not rely on certain default values in thread FPU register, but:
The thread should explicitly set the FPU register again (Set8087CW(), mostly by highlevel functions). 
Then (still after thread reset) the thread-FPU register is safe: Because this thread does not read the value of Default8087CW (possibly changed by other threads).

**Comment by tweibert** [comment]
> But Default8087CW is only read in a very specific situation: when the thread is reset to a default value.

No, it is actually also read when you start a **new** thread, without knowing, unless you happen to look at the procedure `_FpuInit` in `System.pas`, which has the following code line in it:

    Set8087CW(Default8087CW);

This is absolutely unexpected behavior, and not at all documented by EMB. On OS level, and in every other programming language I'm aware of, the FPU flags of a newly created thread will be inherited from the parent thread. Only in Delphi, there is this secret global variable...

When you look at the docs at https://docwiki.embarcadero.com/Libraries/Alexandria/en/System.Default8087CW they just say:

> Default8087CW is the value to which the Floating Point Unit (FPU) control register is set.

Not a single word of warning about the consequences for multi-threaded applications. If they had added something like "It will also be used as the initial control word for new threads", it would have saved my days and weeks of debugging.

I consider this a major design flaw in the RTL, and in order to minimize the impact, I can only advise all developers to avoid these functions whenever they can, or use their own thread-safe versions of the functions that don't mess around with `Default8087CW`, like these: https://gist.github.com/fyndor/4335833f9b7e5418114cdf33cb9b6e43

And it's not as simple as "just set the CW to whatever you want in your own threads" - because there may be threads in (closed source!) 3rd party components. And having a wrong rounding mode is a huge problem, because many functions depend on it, for example `FormatDateTime` (this is how we noticed it in the first place). Want to try for yourself?

````delphi
program test;

{$APPTYPE CONSOLE}

{$R *.res}

uses
  System.SysUtils,
  System.DateUtils,
  System.Math;

var
  dt: TDateTime;

begin
  dt := EncodeTime(16, 00, 00, 000);

  SetRoundMode(rmNearest);
  Writeln(FormatDateTime('hh:nn:ss', dt));  // 16:00:00

  SetRoundMode(rmDown);
  Writeln(FormatDateTime('hh:nn:ss', dt));  // 15:59:59

  Readln;
end.
````

**Comment by AngusJohnson** [comment]
TLDR the whole discussion above 😜, but I think I need to clarify some things:

1. My rasterizing algorithm requires _truncating_ floats to integer values.
2. Previously, `System.Trunc` was too slow (and very much slower that `System.Round`). I'm guessing this was an issue/bug in an older version of Delphi.
3. To address this slow `System.Trunc` issue, I created my own `__Trunc` function which was much faster.
4. IIRC, some time later someone suggested using  `System.Round` together with `System.Math.TRoundingMode` = rmTruncate (default = rmNearest) which was again a little faster still than my own `__Trunc` function.
5. When Torben alerted me to the issue of thread safety in his OP above, it prompted me to reassess `System.Trunc`. Whatever the issue was with this function previously, it seems to have been fixed (in Delphi 11 Upd. 3). So I've reverted to using `System.Trunc` which bypasses any need to alter `System.Math.TRoundingMode` .

**Comment by tomwiel** [comment]
Hm, I thought rmDown (aka floor) was intended (pixel grid independent of sign).

About trunc(), the performance can be fast or slower:
x64:  Delphi uses fast SSE2 (trunc (CVTTSD2SI) without need to change/restore MXCSR) 
x86:  Delphi uses FPU: trunc (changes/restores CW),  round (optimized), the same as in old [guideline](https://www.delphitools.info/OptimalCode/float.htm).

**Comment by AngusJohnson** [comment]
> Hm, I thought rmDown (aka floor) was intended (pixel grid independent of sign).

I don't think it mattered since the raster coordinates were always positive.
